### DISCUSSION TITLE
finding the tableReferenceId

### DISCUSSION THREAD

**Discussion by jmbott** [redivis/meta#188] [Q&A] [answered]
I am looking for the simplest way to find a full `tableIdentifier`. Also to fully understand the `tableReferenceId`

I can see the `tableIdentifier = tableName:tableReferenceId` in the api docs [here](https://apidocs.redivis.com/referencing-resources). I can see in the [definition of a table](https://apidocs.redivis.com/resource-definitions/table) that there is a unique integer referenceId and a name string supplied by the user but I cant find how to display it. 

From the examples I have it seems that the `tableName` is an all lowercase version of the name shown in the user interface with underscores replacing spaces but unsure about the `tableReferenceId`.

Am I missing a display in the user interface for the `tableIdentifier`?
Is there a way to query the `tableIdentifier` through the api? 

I tried the curl example at the bottom of [this](https://apidocs.redivis.com/rest-api/tables/list) page to list the tables. The output seems not to list the info I am looking for. 

I see an example for [reading table data in python](https://apidocs.redivis.com/client-libraries/redivis-python/examples). Here I can query the `table.name` and that follows with the examples I have but when I try `table.identifier` I only see `current` in what I think is the dataset referenceId area (after the `:`) and nothing listed after the table name. 

Is the `tableReferenceId` a version number or unique id? Where should I look for info on it?

**Comment by jmbott** [comment]
[This](https://apidocs.redivis.com/referencing-resources#escaping-names) part of the documentation describes the tableName, 

> 1. All non alpha-numeric and underscore characters in names and version tags are replaced by an underscore (_) character.
> 2. Multiple underscore characters are collapsed into one.
> 3. Leading and trailing underscores are removed.
> 4. All names are case-insensitive. 

That still leaves the reference id though.

**Comment by imathews** [comment, marked as answer]
Hi @jmbott — thanks for opening this. The `referenceId` for a table can be found in the "programmatic" tab of the export interface [example](https://redivis.com/datasets/7br5-41440fjzk/tables?exportTable=6fff-2djw7v7mw&exportDestination=programmatic); make sure to check the box that says "exact references". You'll also see this value gets automatically populated when writing SQL queries in the transform interface, or when creating a new notebook.

Moreover, you can use the URL bar to quickly grab the referenceId; it is always the first part of a table / dataset / project's full id. For example, the URL https://redivis.com/datasets/7br5-41440fjzk/tables/6fff-2djw7v7mw points to dataset with id `7br5-41440fjzk` and table with id `6fff-2djw7v7mw`. The reference id for the dataset is `7br5`, and for the table, `6fff`. This identifier is guaranteed to be unique within the relevant context — that is, there is no other dataset with the id `7br5` within that organization, and no other table with the id `6fff` within that dataset. By using the referenceId within your queries, you ensure they'll continue to work even if the dataset or table gets renamed. 

Finally, this information should be available through the API. It looks like we have a bug where `referenceId` isn't currently populated on the table; I'll get that fixed ASAP. But you can also take the first component of the `id` property, or use the table's `qualifiedReference`, which includes the full referenceId.

Some of the API docs concerning `referenceId` had fallen a bit out of the date; I just went through and tidied up the docs, hopefully it reads clearer now: https://apidocs.redivis.com/referencing-resources

**Comment by jmbott** [comment]
@imathews Thanks for the explination. It makes sense now. 

[This](https://github.com/SEL-Columbia/cwp-epu-data-platform/blob/master/src/config/adminVectors.js#L28) was part of my confusion. The numbers included as `referenceId`s must just be placeholders, meant to be overridden. Unless I am missing something?

From testing it looks like query's still work if either the `referenceId ` or `name` are correct, they fail if both are incorrect, and the `referenceId` overrides the `name`. Does that sound accurate?

**Reply by imathews** [reply]
@jmbott I see yea that code was using a legacy system for the referenceId — I'm sorry, we should've reached out to you when it was fully deprecated. Incremental referenceIds were replaced by a random 4-character sequence, so the correct qualified reference for that table would now be: `modilab.uganda_geodata:2he4.uganda_regions:fzhp`. This approach should be far more future-proof and we don't have any intention to change it again.

Your description of the rules is accurate, with the minor exception that I believe an incorrect four-character referenceId will yield a "not found", even if the name is correct. In this case, because the sequence was only one character, it didn't match the rules for a referenceId and was just ignored.

I'm also going to open an internal issue that it'd be helpful to be able to easily copy this qualified reference from somewhere in the interface, it's a bit of work to find it right now.

**Reply by jmbott** [reply]
Thanks for the explination. As you said nothing seemed to break I just couldnt figure out what should go there when I was adding new layers to the config. This clarifies things.
